pep/003-dependency-namespaces-and-labels.md

```diff
@@ -0,0 +1,508 @@
+* [Solution: Scoped Resource](#solution-scoped-resource)
+* [Solution: User Provided Resource](#solution-user-provided-resource)
+* [Solution: Polymorophic Resource](#solution-polymorphic-resource)
+
+#### Solution: Exclusive Resource
+
+Use a label that would not exist already so that a new installation is created just for this bundle in the current namespace.
+
+Bundle
+```yaml
+dependencies:
+  requires:
+  - name: redis
+    namespace: "{{ installation.namespace }}"
+    labels:
+      owner: "{{ installation.name }}"
+```
+
+
+#### Solution: Global Resource
+
+Install OLM and do not set a namespace. The Porter Operator is installed into a namespace, and it reuses the global installation of OLM. When a match cannot be found, OLM is installed without a namespace set.
+
+```yaml
+dependencies:
+  requires:
+  - name: olm
+```
+
+* When more than one installation matches, installations in the current namespace are preferred over global installations.
+* Note that this looks like today's dependency definition because global is the default right now.
+
+
+#### Solution: Scoped Resource
+
+The micro-services use a namespace criteria for their dependency on postgres. The first micro-service won't find a match, and installs postgres in the current namespace. The subsequent micro-services reuse that installation.
+
+Share the dependency between application instances in the same namespace
+```yaml
+dependencies:
+  requires:
+  - name: postgres
+    namespace: "{{ installation.namespace }}"
+    labels:
+      app: my-application
+```
+
+The labels let us have multiple applications (which can be comprised of micro-services) installed in a namespace and only share a postgres installation between micro-services within the same application.
+
+
+#### Solution: User Provided Resource
+
+SQL Server is either set up by a Porter bundle or manually because it's existing infrastructure. For existing infra, you define a bundle that just exposes the connection string via an output. The bundle itself doesn't modify any resources.
+
+Bundle
+```yaml
+dependencies:
+  requires:
+  - name: sqlserver
+    reference: azure/sqlserver # Use this bundle if we can't find an existing one
+    version: v1.2.x
+
+install:
+- exec:
+    description: "Install something that needs a connection string"
+    commands: ./install.sh
+    arguments:
+    - "{{ bundle.dependencies.sqlserver.outputs.connection-string }}"
+```
+
+Meta Dependency Bundle used to share a resource provisioned outside of Porter. It doesn't do anything but return the connection string given to it when first "installed".
+```yaml
+name: shared-dev-sql-server
+
+parameters:
+- name: connection-string
+  type: string
+  path: /cnab/app/outputs/connection-string
+
+outputs:
+- name: connection-string
+  type: string
+  path: /cnab/app/outputs/connection-string
+```
+
+```console
+$ porter install --reference myapp:v1.0.0 --dependency sqlserver=shared-dev-sql-server
```

Comment:
Question:  If I forget to supply/specify the same `--dependency` on the next action on the installed bundle here (say, upgrade), what should the behavior be?  Perhaps Porter warns, something along the lines of, "It looks like this installation is currently using the specified dependency installation, but this action would require <either the default dependency declared in the bundle, or, if another value was supplied via --dependency, that value>"?

Reply:
I was thinking (but it's not in the PEP yet) is that we need a dependency lockfile that has the resolved graph, including user overrides. So when you override with --dependency, the graph is updated to reflect that. When you do an upgrade, you won't need to re-specify.

Reply:
I'll make sure to transfer this over to the follow-up lifecycle management proposal.